src/acc/calc_dt.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+ Crown Copyright 2012 AWE.
+
+ This file is part of CloverLeaf.
+
+ CloverLeaf is free software: you can redistribute it and/or modify it under
+ the terms of the GNU General Public License as published by the
+ Free Software Foundation, either version 3 of the License, or (at your option)
+ any later version.
+
+ CloverLeaf is distributed in the hope that it will be useful, but
+ WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ details.
+
+ You should have received a copy of the GNU General Public License along with
+ CloverLeaf. If not, see http://www.gnu.org/licenses/.
+ */
+
+#include "calc_dt.h"
+#include <string>
+
+#include <cmath>
+
+//  @brief Fortran timestep kernel
+//  @author Wayne Gaudin
+//  @details Calculates the minimum timestep on the mesh chunk based on the CFL
+//  condition, the velocity gradient and the velocity divergence. A safety
+//  factor is used to ensure numerical stability.
+
+void calc_dt_kernel(bool use_target, int x_min, int x_max, int y_min, int y_max, double dtmin, double dtc_safe, double dtu_safe,
+                    double dtv_safe, double dtdiv_safe, field_type &field, double &dt_min_val, int &dtl_control, double &xl_pos,
+                    double &yl_pos, int &jldt, int &kldt, int &small) {
+
+  small = 0;
+  dt_min_val = g_big;
+  double jk_control = 1.1;
+
+  // DO k=y_min,y_max
+  //   DO j=x_min,x_max
+  //	Kokkos::MDRangePolicy <Kokkos::Rank<2>> policy({x_min + 1, y_min + 1}, {x_max + 2, y_max + 2});
+
+  const int flux_x_stride = field.flux_x_stride;
+  const int flux_y_stride = field.flux_y_stride;
+
+  const int base_stride = field.base_stride;
+  const int vels_wk_stride = field.vels_wk_stride;
+
+  double *xarea = field.xarea.data;
+  double *yarea = field.yarea.data;
+  double *celldx = field.celldx.data;
+  double *celldy = field.celldy.data;
+  double *volume = field.volume.data;
+  double *density0 = field.density0.data;
+  double *viscosity = field.viscosity.data;
+  double *soundspeed = field.soundspeed.data;
+  double *xvel0 = field.xvel0.data;
+  double *yvel0 = field.yvel0.data;
+
+/* TODO I think this loop has room for speedup? Instead of doing 1 do 5 min reductions? */
```

Comment:
You mean `reduction(min: dtct, dtut, dtvt, dtdivt, dt_min_va)` and then a final `fmin(dt_min_val, fmin(dtct, fmin(dtut, fmin(dtvt, fmin(dtdivt, g_big)))));` on the host? I would be interested to see the difference if you have time to benchmark this, preferably using `clover_bm16` or larger, ideally  `clover_bm64` or if memory allows `clover_bm256`.

Reply:
Yes that's what I would like to try. Thanks for the suggestions for the input sizes. I will definitely contact you with any findings I have.